### ISSUE TITLE
Squiz.Arrays.ArrayDeclaration.ValueNoNewline

### ISSUE THREAD

**Issue by evilebottnawi** [squizlabs/PHP_CodeSniffer#715]
Not throw error on this code

``` php
$test = [
    1, 2
];
```

Buth throw error

``` php
$test = [
    1, 2, 3
];
```

**Comment by gsherwood** [comment]
Both arrays throw the same errors. I'm not sure what you are asking or what bug you are reporting. Here is my test:

```
$ phpcs temp.php --standard=Squiz --sniffs=Squiz.Arrays.ArrayDeclaration -s

FILE: temp.php
--------------------------------------------------------------------------------
FOUND 7 ERRORS AFFECTING 4 LINES
--------------------------------------------------------------------------------
 3 | ERROR | [x] Array value not aligned correctly; expected 9 spaces but found
   |       |     4 (Squiz.Arrays.ArrayDeclaration.ValueNotAligned)
 3 | ERROR | [x] Each value in a multi-line array must be on a new line
   |       |     (Squiz.Arrays.ArrayDeclaration.ValueNoNewline)
 4 | ERROR | [x] Closing parenthesis not aligned correctly; expected 8 space(s) but
   |       |     found 0 (Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned)
 7 | ERROR | [x] Array value not aligned correctly; expected 9 spaces but found
   |       |     4 (Squiz.Arrays.ArrayDeclaration.ValueNotAligned)
 7 | ERROR | [x] Each value in a multi-line array must be on a new line
   |       |     (Squiz.Arrays.ArrayDeclaration.ValueNoNewline)
 7 | ERROR | [x] Each value in a multi-line array must be on a new line
   |       |     (Squiz.Arrays.ArrayDeclaration.ValueNoNewline)
 8 | ERROR | [x] Closing parenthesis not aligned correctly; expected 8 space(s) but
   |       |     found 0 (Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned)
--------------------------------------------------------------------------------
PHPCBF CAN FIX THE 7 MARKED SNIFF VIOLATIONS AUTOMATICALLY
--------------------------------------------------------------------------------
```

The first array has one error for the second value not being on a new line. The second array has two errors because the second and third values both need to be on new lines.

**Comment by glen-84** [comment]
@gsherwood Which version are you using? I'm seeing the same issue.

Also, I can repro your output if I add trailing commas. You should remove those, if you added them.

**Comment by gsherwood** [comment]
> Which version are you using?

I was using PHPCS 2.3.4 with PHP 5.5 for that testing (also tried PHP 5.6 and 7), and I'm testing the file as-is (no trailing commas). Here is the file contents:
```php
<?php
$test = [
    1, 2
];

$test = [
    1, 2, 3
];
```

> Edit: FWIW, I also have issues with this:
> 
> $x = [
>     1,
>     2,3
> ];

What issues are you getting? I get the following errors, which seem correct, but is missing a ValueNoNewline (assuming it isn't suppressed because of another issue). Is that what you mean?
```
FILE: temp.php
----------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 3 LINES
----------------------------------------------------------------------
 3 | ERROR | [x] Array value not aligned correctly; expected 6 spaces but
   |       |     found 4 (Squiz.Arrays.ArrayDeclaration.ValueNotAligned)
 4 | ERROR | [x] Array value not aligned correctly; expected 6 spaces but
   |       |     found 4 (Squiz.Arrays.ArrayDeclaration.ValueNotAligned)
 4 | ERROR | [x] Comma required after last value in array declaration
   |       |     (Squiz.Arrays.ArrayDeclaration.NoCommaAfterLast)
 5 | ERROR | [x] Closing parenthesis not aligned correctly; expected 5
   |       |     space(s) but found 0
   |       |     (Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 4 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

But auto-fixing turns it into this, which is what that sniff wants, so the sniff seems to be working fine.
```php
<?php
$x = [
      1,
      2,
      3,
     ];
```

**Comment by glen-84** [comment]
If you don't have trailing commas, then why don't you see `Squiz.Arrays.ArrayDeclaration.NoCommaAfterLast`?

Here is my output (2.3.4 installed using Composer):

    $ ./vendor/bin/phpcs temp.php --standard=Squiz --sniffs=Squiz.Arrays.ArrayDeclaration -s

    FILE: D:\Programming\Projects\x\coding-standard-php\temp.php
    ----------------------------------------------------------------------
    FOUND 7 ERRORS AFFECTING 4 LINES
    ----------------------------------------------------------------------
     3 | ERROR | [x] Array value not aligned correctly; expected 9 spaces but
       |       |     found 4 (Squiz.Arrays.ArrayDeclaration.ValueNotAligned)
     3 | ERROR | [x] Comma required after last value in array declaration
       |       |     (Squiz.Arrays.ArrayDeclaration.NoCommaAfterLast)
     4 | ERROR | [x] Closing parenthesis not aligned correctly; expected 8
       |       |     space(s) but found 0
       |       |     (Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned)
     7 | ERROR | [x] Array value not aligned correctly; expected 9 spaces but
       |       |     found 4 (Squiz.Arrays.ArrayDeclaration.ValueNotAligned)
     7 | ERROR | [x] Each value in a multi-line array must be on a new
       |       |     line (Squiz.Arrays.ArrayDeclaration.ValueNoNewline)
     7 | ERROR | [x] Comma required after last value in array declaration
       |       |     (Squiz.Arrays.ArrayDeclaration.NoCommaAfterLast)
     8 | ERROR | [x] Closing parenthesis not aligned correctly; expected 8
       |       |     space(s) but found 0
       |       |     (Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned)
    ----------------------------------------------------------------------
    PHPCBF CAN FIX THE 7 MARKED SNIFF VIOLATIONS AUTOMATICALLY
    ----------------------------------------------------------------------

    Time: 76ms; Memory: 2.25Mb

> What issues are you getting? I get the following errors, which seem correct, but is missing a ValueNoNewline (assuming it isn't suppressed because of another issue). Is that what you mean?

Yes, the missing `ValueNoNewline`.

**Comment by gsherwood** [comment]
> If you don't have trailing commas, then why don't you see Squiz.Arrays.ArrayDeclaration.NoCommaAfterLast?

Great question. I ran the command again and now I see it just fine. I get the same results as you. So I obviously stuffed something up when I originally ran it.

**Comment by gsherwood** [comment]
The reason you are not getting a `ValueNotNewline` on the first array is because there is no trailing comma. The way the sniff determines the array indexes is by looking for those commas and processing the values before them. If the last value doesn't have a comma, it can't see that last value and so generates a special `NoCommaAfterLast ` error for it.

Once you fix that error, the sniff can process the last array value correctly and produce new errors for you. If you are using auto-fixing with `phpcbf` then this process happens automatically as the fixer loops to find additional fixes, as you can see via the output from `--report=diff -vv`:
```
*** START FILE FIXING ***
	E: [Line 4] Closing parenthesis not aligned correctly; expected 8 space(s) but found 0 (Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned)
	Squiz_Sniffs_Arrays_ArrayDeclarationSniff (line 340) replaced token 12 (T_WHITESPACE) "\n]" => "\n········]"
	E: [Line 3] Comma required after last value in array declaration (Squiz.Arrays.ArrayDeclaration.NoCommaAfterLast)
	Squiz_Sniffs_Arrays_ArrayDeclarationSniff (line 608) replaced token 11 (T_LNUMBER) "2" => "2,"
	E: [Line 3] Array value not aligned correctly; expected 9 spaces but found 4 (Squiz.Arrays.ArrayDeclaration.ValueNotAligned)
	Squiz_Sniffs_Arrays_ArrayDeclarationSniff (line 650) replaced token 7 (T_WHITESPACE) "····1" => "·········1"
	* fixed 3 violations, starting loop 2 *
	E: [Line 3] Each value in a multi-line array must be on a new line (Squiz.Arrays.ArrayDeclaration.ValueNoNewline)
	Squiz_Sniffs_Arrays_ArrayDeclarationSniff (line 628) replaced token 10 (T_WHITESPACE) "·2" => "2"
	Squiz_Sniffs_Arrays_ArrayDeclarationSniff (line 631) replaced token 11 (T_LNUMBER) "2" => "\n2"
	* fixed 2 violations, starting loop 3 *
	E: [Line 4] Array value not aligned correctly; expected 9 spaces but found 0 (Squiz.Arrays.ArrayDeclaration.ValueNotAligned)
	Squiz_Sniffs_Arrays_ArrayDeclarationSniff (line 648) replaced token 10 (T_WHITESPACE) "\n2" => "\n·········2"
	* fixed 1 violations, starting loop 4 *
	*** END FILE FIXING ***
```

If I change the way the sniff processes the array, I can try and get it to process that final value correctly and produce more error up front. But right now, it's working fine (and the same way it always has) so I'll leave this as an enhancement. It used to come up more often, so I wrote an FAQ about it many many years ago: https://github.com/squizlabs/PHP_CodeSniffer/wiki/FAQ#how-come-php_codesniffer-reported-errors-i-fixed-them-now-i-get-even-more

**Comment by gsherwood** [comment]
Closing as this change isn't going to be made after so many years.
